### PR TITLE
Add folding

### PR DESCRIPTION
**Pull request by xiaoxiaowesley**
This PR is for adding a folding feature.

- support block code folding.
- support collection code folding

Thank you for reviewing!

### CODE REVIEW

File: src/main/java/com/google/idea/gn/GnFoldingBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.PsiTreeUtil
+import java.util.*
+
+class GnFoldingBuilder : FoldingBuilderEx(), DumbAware {
+    override fun buildFoldRegions(root: PsiElement, document: Document, quick: Boolean): Array<FoldingDescriptor> {
+        val descriptors: MutableList<FoldingDescriptor> = ArrayList()
+
+        //Block
+        val blocks = PsiTreeUtil.findChildrenOfType<GnBlockImpl>(root, GnBlockImpl::class.java)
```
Comment (bc-lee) [inline src/main/java/com/google/idea/gn/GnFoldingBuilder.kt:20]: nit: Explicit type arguments can be omitted.

File: src/main/java/com/google/idea/gn/GnFoldingBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            val first = block.firstChild
+            val last = block.lastChild
+            val newRange = TextRange(first.textRange.endOffset, last.textRange.startOffset)
+            if (newRange.length > 0) {
+                descriptors.add(FoldingDescriptor(block, newRange))
+            }
+        }
+
+        // Collections
+        val collections = PsiTreeUtil.findChildrenOfType<GnCollectionImpl>(root, GnCollectionImpl::class.java)
```
Comment (bc-lee) [inline src/main/java/com/google/idea/gn/GnFoldingBuilder.kt:31]: This, too.

File: src/test/java/com/google/idea/gn/FoldingTest.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package com.google.idea.gn
+
+import com.google.idea.gn.util.GnCodeInsightTestCase
+import org.junit.Test
+
+open class FoldingTest : GnCodeInsightTestCase() {
+    @Test
+    fun testBlockFolding() {
+        var filePath = "$testDataPath/project/src/test/FoldingTest.gn"
+        myFixture.testFolding(filePath);
```
Comment (bc-lee) [inline src/test/java/com/google/idea/gn/FoldingTest.kt:10]: nit: Semicolons are not required.

**Review by bc-lee — COMMENTED** [review]
Thanks for the patch! Looks awesome!

File: src/main/java/com/google/idea/gn/GnFoldingBuilder.kt (hunk 1 above)
Comment (xiaoxiaowesley) [reply src/main/java/com/google/idea/gn/GnFoldingBuilder.kt]: Explicit type arguments have been removed. 👌

File: src/test/java/com/google/idea/gn/FoldingTest.kt (hunk 3 above)
Comment (xiaoxiaowesley) [reply src/test/java/com/google/idea/gn/FoldingTest.kt]: done.

**Review by bc-lee — APPROVED** [review]
Thanks!

**Comment by xiaoxiaowesley** [comment]
@bc-lee This feature hasn't been uploaded to JB Marketplace. Would you help to upload it? Thank you!

**Comment by bc-lee** [comment]
Thanks for the reminder. I've uploaded the updated plugin now, so it should be distributed within 2 days.

**Comment by bc-lee** [comment]
The updated version is now distributed on Jetbrains Marketplace. :)

**Pull request merged**